indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,549 @@
+                                      .addData("task", task.getId())
+                                      .emit();
+                              }
+                            }
+                            catch (Exception e) {
+                              LOGGER.error(e, "Suppressing exception caught while cleaning up task");
+                            }
+
+                            // Make sure we clear the interrupt flag at the end, since this thread will be reused
+                            // for other tasks.
+                            Thread.interrupted();
+                          }
+                        }
+                      }
+                  )
+              )
+      );
+      saveRunningTasks();
+      return tasks.get(task.getId()).getResult();
+    }
+  }
+
+  @Override
+  public void shutdown(String taskid, String reason)
+  {
+    LOGGER.info("Shutdown [%s] because: [%s]", taskid, reason);
+    final ThreadingTaskRunnerWorkItem taskInfo;
+
+    synchronized (tasks) {
+      taskInfo = tasks.get(taskid);
+
+      if (taskInfo == null) {
+        LOGGER.info("Ignoring request to cancel unknown task: %s", taskid);
+        return;
+      }
+
+      if (taskInfo.shutdown) {
+        LOGGER.info(
+            "Task [%s] is already shutting down, ignoring duplicate shutdown request with reason [%s]",
+            taskid,
+            reason
+        );
+      } else {
+        taskInfo.shutdown = true;
+        scheduleTaskShutdown(taskInfo);
+      }
+    }
+  }
+
+  /**
+   * Submits a callable to the control thread pool that attempts a task graceful shutdown,
+   * if shutdown is not already scheduled.
+   *
+   * The shutdown will wait for the configured timeout and then interrupt the thread if the timeout is exceeded.
+   */
+  private ListenableFuture scheduleTaskShutdown(ThreadingTaskRunnerWorkItem taskInfo)
+  {
+    synchronized (tasks) {
+      if (taskInfo.shutdownFuture != null) {
+        return taskInfo.shutdownFuture;
+      }
+
+      taskInfo.shutdownFuture = controlThreadExecutor.submit(
+          new Callable<Void>()
+          {
+            @Override
+            public Void call()
+            {
+              LOGGER.info("Stopping thread for task: %s", taskInfo.getTaskId());
+              taskInfo.getTask().stopGracefully(taskConfig);
+
+              try {
+                TaskStatus status = taskInfo.getResult().get(
+                    taskConfig.getGracefulShutdownTimeout().toStandardDuration().getMillis(),
+                    TimeUnit.MILLISECONDS
+                );
+
+                if (status == null) {
+                  // Note that we can't truly force a hard termination of the task thread externally.
+                  // In the future we may want to add a forceful shutdown method to the Task interface.
+                  if (taskInfo.shutdownFuture != null) {
+                    taskInfo.shutdownFuture.cancel(true);
```

Comment:
we need to cancel status future and not shutdownFuture.
also Future.get(..) throws TimeoutException instead of null, so this will not be called if task didn't finish within timeout
also, we can remove keeping the thread as "state"

anyways, I thought it would be easier to write code :) , can you merge https://github.com/jon-wei/druid/pull/4 ?

Reply:
Ah, my bad, cancelled the wrong future.

Thanks for the patch!